graphql_compiler/interpreter/debugging.py

```diff
@@ -0,0 +1,195 @@
+from copy import deepcopy
+from dataclasses import dataclass, field
+from typing import Any, ClassVar, Dict, Generic, Iterable, List, Sequence, Tuple, TypeVar
+
+from ..typedefs import Literal
+from .typedefs import DataToken
+
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def _enumerate_starting_at(
+    sequence: Sequence[T],
+    first_index: int,
+) -> Iterable[Tuple[int, T]]:
+    """Enumerate only a portion of a sequence, without slicing the sequence.
+
+    Equivalent to enumerate(sequence[first_index:], start=first_index). The built-in enumerate()
+    always iterates the entire sequence and only changes the counter's value based on
+    the start parameter. In this function, we skip elements with indices in [0, first_index)
+    in the sequence, and only iterate over elements with indices in [first_index, len(sequence)).
+
+    Args:
+        sequence: arbitrary sequence of values over which to iterate
+        first_index: the first index of the sequence to produce in the resulting iterable
+
+    Yields:
+        tuples (index, element_at_that_index_in_sequence), analogous to the built-in enumerate()
+    """
+    for index in range(first_index, len(sequence)):
+        yield index, sequence[index]
+
+
+def _unzip_and_yield_second(iterable: Iterable[Tuple[T, U]]) -> Iterable[U]:
+    """Unpack an iterable of tuples and only yield the second element of each tuple."""
+    for _, second in iterable:
+        yield second
+
+
+@dataclass(frozen=True)
+class AdapterOperation:
+    """The record of an action performed on or by an InterpreterAdapter function as part of a trace.
+
+    For example, calling get_tokens_of_type("Foo") would produce an AdapterOperation
+    with kind "call", with name "get_tokens_of_type", with unique ID and parent unique IDs
+    appropriate for the trace, and with data (("Foo",), {}) as the tuple of
+    the positional arguments tuple and kwargs dict.
+
+    The parent_uid field allows us to track dependencies across different operations. For example,
+    data yielded by a generator-style function will be recorded in AdapterOperations
+    of kind "yield" whose parent_uids point to the AdapterOperation of the particular invocation
+    of the function that is generating that data.
+
+    When InterpreterAdapter functions consume iterable arguments, the iterable argument in
+    the AdapterOperation "call" record is replaced by a placeholder name. As values from such
+    an iterable are consumed, those values are recorded in their own AdapterOperation records
+    with kind "yield", parent unique ID matching the unique ID of the "call" record, and
+    name matching the placeholder name in the "call" record.
+    """
+
+    kind: Literal["call", "yield", "return"]
+    name: str
+    uid: int
+    parent_uid: int
+    data: Any
```

Comment:
This sentence: `data (("Foo",), {}) as the tuple of the positional arguments tuple and kwargs dict.` makes `data` sound like something specific so being type `Any` confused me. Looks it varies based on the `kind`? I think more explanation is need for what `data` is expected to be.

Reply:
I added some more information about what `data` is expected to be. Unfortunately, when recording a `kind="yield"` operation, the `data` value can be a variety of different things that are specific to the use case and recorder class/function being used. I tried to clarify that as best I could while still trying to keep the docstring concise, let me know what you think.